mobile-browser-based-version/src/helpers/model_definition/model.js

```diff
@@ -0,0 +1,180 @@
+import * as tf from '@tensorflow/tfjs';
+import { InteroperabilityLayer } from '../model_definition/custom_layers';
+import { getWorkingModel, getWorkingModelMetadata } from '../memory/helpers';
+
+/**
+ * Enumeration of the different types of personalization.
+ */
+export const personalizationType = {
+  NONE: 'NONE',
+  INTEROPERABILITY: 'interoperability',
+};
+
+/**
+ * This class represents a default model and will simply
+ * encapsulate the default model for a given task.
+ *
+ * In order to create a new Personalized model, you should only need to redefine _createModel() and getSharedModel().
+ */
+export class Model {
+  constructor(task, useIndexedDB) {
+    this.task = task;
+    this.dataType = task.trainingInformation.dataType;
+    this.useIndexedDB = useIndexedDB;
+    this.typeOfPersonalization = personalizationType.NONE;
+  }
+
+  /**
+   * Initializes the model for a given task.
+   *
+   * It will check the metadatas we stored to see if there is already asuch model in memory and if so returns it.
+   *
+   * Otherwise (no such model or we are not using index DB) it will create a new model.
+   *
+   * Finally if the type of the previously stored model doesn't correspond to the one we are creating, in will also create a new one.
+   *
+   * This method should not need to be overriden by subclasses.
+   */
+  async init() {
+    let model;
+    let modelParams = [this.task.taskID, this.task.trainingInformation.modelID];
+
+    // Check if a working model already exists in storage.
+    if (this.useIndexedDB && (await getWorkingModelMetadata(...modelParams))) {
+      console.log('Loading model from memory');
+      model = await getWorkingModel(...modelParams);
+    } else {
+      console.log(`Creating new ${this.typeOfPersonalization} model`);
+      model = await this._createModel();
+    }
+
+    // Make sure that the current working model is indeed the type we expect.
+    if (this.isOfCorrectType(model)) {
+      this.model = model;
+    } else {
+      console.log(
+        `Previous working model was not of type ${this.typeOfPersonalization} : creating a new one`
+      );
+      this.model = await this._createModel();
+    }
+  }
+
+  /**
+   * Private method to create the model.
+   *
+   * This should be overriden in sublcasses in order to define the personalized model.
+   * @returns the newly created model.
+   */
+  async _createModel() {
+    return await this.task.createModel();
+  }
+
+  /**
+   *  Getter for the model.
+   *  @returns the model we want to train.
+   */
+  getModel() {
+    return this.model;
+  }
+
+  /**
+   *  This is a getter for the part of the model we want to share with the network.
+   *  For personalizationType.NONE this is simply the model.
+   *  The behaviour depends on the personalization type of the model.
+   *  @returns the part of the model we want to share.
+   */
+  getSharedModel() {
+    return this.model;
```

Comment:
It's confusing that both getModel and getSharedModel return the same object, when we want to use it. for someone new it might not be clear which to use. Maybe we should only keep getSharedModel? What is the reason to have 2, just for naming?

Reply:
Yeah that's the thing, this only the case for personalizationType.NONE. because here both the local and shared model are the model itself. You actually change this when redefining a new subclass.